Reliable persistence

v0d1ch opened the pull request:
handles part of #1079 

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Pass in persistence handle and append on broadcast

v0d1ch committed:
Make everything compile using mockPersistence

v0d1ch committed:
Add a test to assert we append messages in reliability layer

v0d1ch committed:
Check for the existence of the persistence file as well

v0d1ch committed:
Save vector clock messages too

v0d1ch committed:
Use own mock persistence for messages

ffakenz commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -133,19 +134,28 @@ instance Arbitrary ReliabilityLog where
   Tracer m ReliabilityLog ->
+  -- | Persistence handle to store messages
+  PersistenceIncremental (Heartbeat msg) m ->
+  -- | Persistence handle to store acks
```
```suggestion
  -- | Persistence handle to store ack counters (clocks vector).
```

ffakenz commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -133,19 +134,28 @@ instance Arbitrary ReliabilityLog where
+  (MonadThrow (STM m), MonadThrow m, MonadAsync m, ToJSON msg, FromJSON msg) =>
   -- | Tracer for logging messages.
   Tracer m ReliabilityLog ->
+  -- | Persistence handle to store messages
```
```suggestion
  -- | Persistence handle to store sent network messages.
```

ffakenz commented on hydra-node/src/Hydra/Node/Network.hs:
```diff
@@ -74,15 +74,21 @@ import Hydra.Network.Heartbeat (ConnectionMessages, Heartbeat (..), withHeartbea
+  (ToCBOR msg, ToJSON msg, FromJSON msg, FromCBOR msg) =>
   -- | Tracer to use for logging messages.
   Tracer IO (LogEntry tx msg) ->
+  -- | Persistence handle to store messages
```
```suggestion
  -- | Persistence handle to store sent network messages.
```

ffakenz commented on hydra-node/src/Hydra/Node/Network.hs:
```diff
@@ -74,15 +74,21 @@ import Hydra.Network.Heartbeat (ConnectionMessages, Heartbeat (..), withHeartbea
   Tracer IO (LogEntry tx msg) ->
+  -- | Persistence handle to store messages
+  PersistenceIncremental (Heartbeat msg) IO ->
+  -- | Persistence handle to store acks
```
```suggestion
  -- | Persistence handle to store ack counters (clocks vector).
```

v0d1ch replied:
Actually tt's called vector clock.

ffakenz commented on hydra-node/test/Hydra/Network/ReliabilitySpec.hs:
```diff
@@ -151,6 +158,40 @@ spec = parallel $ do
       receivedMsgs `shouldBe` [ReliableMsg (fromList [1, 1]) (Data "node-1" msg)]
+
+    it "appends messages to disk and can load them back" $ do
+      withTempDir "" $ \tmpDir -> do
```
```suggestion
      withTempDir "network-messages-persistence" $ \tmpDir -> do
```

ffakenz committed:
Minor formatting changes to network reliability persistence spec

v0d1ch committed:
Remove sent messages from memory

v0d1ch committed:
Rename mock message handles in ReliabilitySpec

v0d1ch committed:
Introduce MessagePersistence handle

abailly committed:
Store messages in right order with mock persistence

v0d1ch committed:
Finalize MessagePersistence handle

v0d1ch committed:
Remove unneeded TVars in favor of persistence handle

v0d1ch committed:
Improve log message in case of missing local message to resend

v0d1ch committed:
Create persistence handle inside of network code

abailly committed:
Fix concurrent update of vector clocks

We removed the use of TVar but callback is not thread safe and
therefore we need to ensure updates to the acks are atomic across
broadcast and callback calls, which is not possible when writing to
disk in both calls.

Broadcast is "guaranteed" to be called from a single thread and
therefore it's the only place where we can safely write to disk.

abailly committed:
Simplify creation of Message persistence handle

v0d1ch committed:
Go back to using me + others interface

v0d1ch committed:
Go back to using me + others interface

v0d1ch committed:
Add ReceivedMalformedAcks trace

abailly committed:
Update module documentation

abailly committed:
Save vector clock also on Pings
